crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1051,15 +1051,22 @@ impl<'db> Type<'db> {
 
             (Type::SubclassOf(_), Type::SubclassOf(_)) => false,
 
-            (Type::SubclassOf(_), Type::Instance(_)) | (Type::Instance(_), Type::SubclassOf(_)) => {
-                false
+            (Type::SubclassOf(_), Type::Instance(instance))
+            | (Type::Instance(instance), Type::SubclassOf(_)) => {
+                // TODO this should be `true` if the instance is of a final type which is not a
+                // subclass of type. (With non-final types, we never know whether a subclass might
+                // multiply-inherit `type` or a subclass of it, and thus not be disjoint with
+                // `type[...]`.) Until we support finality, hardcode None, which is known to be
+                // final.
+                matches!(instance.class.known(db), Some(KnownClass::NoneType))
             }
```

Comment:
We could generalise this branch a _little_ bit to cover all singleton classes that are known not to be subtypes of `type`:

```diff
--- a/crates/red_knot_python_semantic/src/types.rs
+++ b/crates/red_knot_python_semantic/src/types.rs
@@ -1051,14 +1051,15 @@ impl<'db> Type<'db> {
 
             (Type::SubclassOf(_), Type::SubclassOf(_)) => false,
 
-            (Type::SubclassOf(_), Type::Instance(instance))
-            | (Type::Instance(instance), Type::SubclassOf(_)) => {
+            (Type::SubclassOf(_), instance @ Type::Instance(InstanceType { class }))
+            | (instance @ Type::Instance(InstanceType { class }), Type::SubclassOf(_)) => {
                 // TODO this should be `true` if the instance is of a final type which is not a
                 // subclass of type. (With non-final types, we never know whether a subclass might
                 // multiply-inherit `type` or a subclass of it, and thus not be disjoint with
-                // `type[...]`.) Until we support finality, hardcode None, which is known to be
-                // final.
-                matches!(instance.class.known(db), Some(KnownClass::NoneType))
+                // `type[...]`.) Until we support finality, special-case various classes
+                // known to be singletons
+                class.known(db).is_some_and(KnownClass::is_singleton)
+                    && !instance.is_subtype_of(db, KnownClass::Type.to_instance(db))
             }
 
             (
```